portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java

```diff
@@ -259,6 +292,26 @@ else if (_printWriter != null) {
 		}
 	}
 
+	private String _addClosingTags(String xml) {
```

Comment:
@ngriffin7a these regex replacements are very slow, please change them to plain string manipulation and resend, thanks!

Reply:
@shuyangzhou The slower regex replacements only execute if for example `if (xml.contains("<link")) {` returns true. I think the elegance of the regex is very nice, and the replacement only operates on a single line of text. Given those points, would you still like me to refactor or can we stay with the regex?

Reply:
@ngriffin7a can you add a new commit to remove the regex? I will see how bad it is after, it is too complex, we can stay with regex. But I prefer no regex if we can.